code/modules/mob/living/carbon/xenomorph/castes/hivemind/hivemind.dm

```diff
@@ -170,8 +188,90 @@
 	var/health_percent = round((max_integrity / obj_integrity) * 100)
 	switch(health_percent)
 		if(-INFINITY to 25)
-			to_chat(parent, "<span class='xenohighdanger'>Your core is under attack, and dangerous low on health!</span>")
+			to_chat(parent, "<span class='xenohighdanger'>Your core is under attack, and is dangerously low on health!</span>")
 		if(26 to 75)
-			to_chat(parent, "<span class='xenodanger'>Your core is under attack, and low on health!</span>")
+			to_chat(parent, "<span class='xenodanger'>Your core is under attack, and is low on health!</span>")
 		if(76 to INFINITY)
 			to_chat(parent, "<span class='xenodanger'>Your core is under attack!</span>")
+
+/mob/living/carbon/xenomorph/hivemind/upgrade_xeno(newlevel) //where we also upgrade the max health of our core and restore it to maximum health
+	. = ..()
+	to_chat(src, "<span class='xenonotice'>Our core strengthens, and becomes more resilient!</span>")
+	core.max_integrity = src.xeno_caste.core_maximum_hitpoints
+	core.obj_integrity = core.max_integrity
+
+/mob/living/carbon/xenomorph/proc/hivemind_weed_vibecheck()
+	if(!src) //Sanity
+		return FALSE
+	if(!istype(src, /mob/living/carbon/xenomorph/hivemind)) //Not a hivemind, we don't care
+		return TRUE
+	var/mob/living/carbon/xenomorph/hivemind/H = src
+	if(!H.core) //Sanity check for a core.
+		return FALSE
+	if(locate(/obj/effect/alien/weeds) in range(1, src.loc))
+		return TRUE
+	H.forceMove(get_turf(H.core))
+	to_chat(src, "<span class='xenonotice'>We lacked weeds to sustain our presence and our consciousness was shunted to our core.</span>")
+	return FALSE
+
+/mob/living/carbon/xenomorph/hivemind/proc/hivemind_core_alert()
+
+	if(!src) //Sanity check
+		return FALSE
+
+	var/mob/living/carbon/xenomorph/hivemind/X = src
+	var/obj/effect/alien/hivemindcore/core = X.core
+
+	if(!core) //Sanity check
+		return FALSE
+
+	var/list/decoy_area_list = list()
+	var/area/real_area = get_area(core) //Set our real area
+	var/list/buffer_list = list() //Buffer list for randomization
+	var/list/details = list() //The actual final list for the announcement
+
+	for(var/area/core_areas in world) //Build the list of areas on the core's Z.
+		if(core_areas.z != core.z) //Must be on the same Z
+			continue
+		decoy_area_list += core_areas //Add to the list of potential decoys
```

Comment:
For the love of god never EVER EVER use for(var in world), use glob lists for what you need instead

Reply:
Do you know of a glob list I can reference?